Framework/Geometry/src/Objects/CSGObject.cpp

```diff
@@ -1086,17 +1086,53 @@ int CSGObject::interceptSurface(Geometry::Track &track) const {
   const auto &IPoints(LI.getPoints());
   const auto &dPoints(LI.getDistance());
 
-  auto ditr = dPoints.begin();
-  auto itrEnd = IPoints.end();
-  for (auto iitr = IPoints.begin(); iitr != itrEnd; ++iitr, ++ditr) {
-    if (*ditr > 0.0) // only interested in forward going points
-    {
-      // Is the point and enterance/exit Point
-      const TrackDirection flag = calcValidType(*iitr, track.direction());
-      if (flag != TrackDirection::INVALID)
-        track.addPoint(flag, *iitr, *this);
+  // sort the points based on its
+  // 1. build a vector that contains the relative distance to the starting point
+  //    of the track/ray
+  // 2. get the index in the sorted order
+  const size_t nPoints(IPoints.size());
+  std::vector<size_t> idxs(nPoints, 0);
+  if (nPoints > 1) {
+    std::vector<double> dists(nPoints);
+    for (size_t i = 0; i < nPoints; i++) {
+      dists[i] = IPoints[i].distance(track.startPoint());
     }
+    // sort by distance, only keep index
+    std::iota(idxs.begin(), idxs.end(), 0);
+    std::sort(idxs.begin(), idxs.end(), [&dists](size_t a, size_t b) { return dists[a] < dists[b]; });
   }
+
+  // 3. iterating through IPoints by the order of the smallest
+  //    relative distance to the largest
+  for (size_t i = 0; i < nPoints; ++i) {
+    const size_t idx(idxs[i]);
+    const auto &currentPt(IPoints[idx]);
+    const auto &prePoint = (i == 0) ? track.startPoint() : IPoints[idxs[i - 1]];
+    const auto ditr = dPoints[idx];
+    if (ditr > 0) {
+      // figuring out the track type (invalid, entering or leaving) by checking
+      // previous point, current point and the downstream point of the current
+      // point.
+      const TrackDirection trackType = calcValidTypeByMidPoint(prePoint, currentPt, track.direction());
+      // add to list
+      track.addPoint(trackType, currentPt, *this);
```

Comment:
This call to addPoint used to have an if statement around it:
```
if (flag != TrackDirection::INVALID)
```
Is that still needed? I think yes

Reply:
Nice catch, thanks.  This should helps with some of the failing unit tests :smile: